File: examples/experiments/multilang/nodejs/memory_rw.js
function intParam(params, name, defaultValue, min, max) {
    const parsed = Number.parseInt(params?.[name] ?? defaultValue, 10);
    const value = Number.isFinite(parsed) ? parsed : defaultValue;
    return Math.max(min, Math.min(max, value));
}

module.exports = (params, ctx) => {
    params = params || {};

    const elements = intParam(
        params,
        "elements",
        3000000,
        100000,
        4000000
    );

    const rounds = intParam(
        params,
        "rounds",
        40,
        1,
        200
    );

    const data = new Uint32Array(elements);

    for (let i = 0; i < elements; i++) {
        data[i] =
            Math.imul(i, 0x9e3779b1) >>> 0;
    }

    let checksum = 0;

    for (let r = 0; r < rounds; r++) {
        const step = 97 + (r % 13);
        const offset = (r * 8191) % elements;

        for (
            let i = 0;
            i < elements;
            i += step
        ) {
            const idx =
                (i + offset) % elements;

            let value = data[idx];

            value = (
                value
                ^ Math.imul(
                    idx + r,
                    0x85ebca77
                )
            ) >>> 0;

            data[idx] = value;

            checksum = (
                checksum + value
            ) >>> 0;
        }

        let local = 0;

        for (
            let i = 0;
            i < elements;
            i += 4
        ) {
            local = (
                local + data[i]
            ) >>> 0;
        }

        checksum = (
            checksum ^ local
        ) >>> 0;
    }

    return {
        benchmark: "memory_rw",
        elements,
        rounds,
        bytes:
            elements
            * Uint32Array.BYTES_PER_ELEMENT,
        checksum,
    };
};

File: examples/experiments/multilang/python/memory_rw.py
from array import array


def _int_param(params, name, default, minimum, maximum):
    try:
        value = int(params.get(name, default))
    except (TypeError, ValueError, AttributeError):
        value = default
    return max(minimum, min(maximum, value))


def handler(params, context):
    params = params or {}

    elements = _int_param(
        params,
        "elements",
        3_000_000,
        100_000,
        4_000_000,
    )

    rounds = _int_param(
        params,
        "rounds",
        40,
        1,
        200,
    )

    data = array(
        "I",
        (
            (i * 2654435761) & 0xFFFFFFFF
            for i in range(elements)
        ),
    )

    mask = 0xFFFFFFFF
    checksum = 0

    for r in range(rounds):
        step = 97 + (r % 13)
        offset = (r * 8191) % elements

        for i in range(0, elements, step):
            idx = (i + offset) % elements
            value = data[idx]

            value = (
                            value
                            ^ ((idx + r) * 2246822519)
                    ) & mask

            data[idx] = value

            checksum = (
                               checksum + value
                       ) & mask

        local = 0

        for i in range(0, elements, 4):
            local = (
                            local + data[i]
                    ) & mask

        checksum ^= local

    return {
        "benchmark": "memory_rw",
        "elements": elements,
        "rounds": rounds,
        "bytes": elements * data.itemsize,
        "checksum": checksum,
    }